### PR TITLE
Implement shortened methods with => syntax

### PR DESCRIPTION
**Pull request by adamdruppe**
For function literals, D offers a short syntax: `x => y`,
which expands to `function (x) { return y; }`. However, for
normal function definitions, there was no such shortening.
For various applications, there can be more syntax than
meaning, such as property accessors.

This commit changes that by expanding the same `=>` syntax
we already have to work in declarations as well. It expands
to the same thing: `int foo() => x;` is simply
`int foo() { return x; }`; this is just shortened syntax.
Combined with existing rules like auto returns, a property
getter can be be as simple as `@property x() => x_;` or a
range-based pipeline may appear like
`auto common_operation() => this[].sort.uniq;` giving D's
existing functional strengths more syntax sugar.

C# has demonstrated the utility of such shortened methods
since its version 7. However, while C# allows it for
constructors as well, this commit will parse it but fail
in semantic because you cannot return a value from a
constructor. I am not convinced it is worth special-casing
the simple "=> y ALWAYS means "{ return y; }" rule for
this case and thus left it alone.

### CODE REVIEW

**Comment by 12345swordy** [comment]
You need a preview switch for this.

**Comment by thewilsonator** [comment]
>this commit will parse it but fail in semantic because you cannot return a value from a constructor

Is it possible to detect this and give a nice error message?

**Comment by adamdruppe** [comment]
No need; there already is one:

`class A {   this() => 7; }`

cool.d(2): Error: cannot return expression from constructor

File: src/dmd/parse.d
Comment (adamdruppe) [inline src/dmd/parse.d:5091]: I think I did this wrong. I just copied from below but p sure this is specifically for }.... should I instead set it to `token.loc` for the semicolon?

Comment (thewilsonator) [reply src/dmd/parse.d:5091]: Seems reasonable to me, and I can't think of any more appropriate token to point it at. It seems to be used only for line mappings for debug info (and I supposed DMD as a Library for whatever).

You should probably update [this comment](https://github.com/dlang/dmd/blob/1339fb58c007ac20eaaeff9549607f9b593a98a6/src/dmd/parse.d#L288) too.

**Comment by thewilsonator** [comment]
Have you made a forum post about this yet? If so please link.

**Comment by UplinkCoder** [comment]
:shrug:

**Comment by Geod24** [comment]
Been wanting this a few times, so obviously 👍 

Few things:
- Are there any context where the syntax could be ambiguous ? `auto` comes to mind (delegate member variable vs function).
- Need spec PR, obviously;
- I'd also open a forum thread, as discussed;
- This *probably* needs either a `-preview` or a DIP;
- Does this work with mixins ?

**Comment by UplinkCoder** [comment]
It's a visible language change.
Needs a DIP.

**Comment by adamdruppe** [comment]
On Thu, Oct 08, 2020 at 09:14:47PM -0700, Nicholas Wilson wrote:
> Have you made a forum post about this yet? If so please link.

No, this was just from a short chat on the discord. People were talking
about the idea, I said that should be easy to implement, they said "lol
prove it" and i was like "lol k".

We can make one in a bit though.

**Comment by adamdruppe** [comment]
On Fri, Oct 09, 2020 at 04:34:15AM -0700, Mathias LANG wrote:
> - Are there any context where the syntax could be ambiguous ? `auto` comes to mind (delegate member variable vs function).

I very much doubt it; it should have failed the tests if it was.

The key difference is the parens.

auto a = () => 4; // parens on rhs of =, a is a variable assigned a
delegate

auto c() => 5; // parens on lhs of =, c is a function

The parens after is what the parser looks for to determine it is a
declaration which is then on a different branch than expression parsing.

Since the latter is always illegal in today's D, it is a purely additive
change and should break nothing. The auto tester seems to agree, aside
from what appears to be spurious nonsense in the (Windows_LDC_Debug
x64-debug-ldc)  thing...

> - Does this work with mixins ?

Can't imagine why it wouldn't. A quick test of

        mixin("auto c() => 5;");

seems to agree.

File: src/dmd/parse.d
Comment (adamdruppe) [reply src/dmd/parse.d:5091]: That comment is the one on the parser class, this is on the f object which is the FuncDeclaration class.  (It was that comment that tipped me I was doing the wrong thing though.)

**Comment by adamdruppe** [comment]
anyone what's up with the azure-pipelines Azure pipelines (Windows_LDC_MinGW win32-ldc)  test?

**Comment by thewilsonator** [comment]
I've seen it on a few other PRs.

**Comment by adamdruppe** [comment]
Forum post up: https://forum.dlang.org/post/cmgqwabzdqbtngmjidfw@forum.dlang.org

**Comment by UplinkCoder** [comment]
Druntime for mingw printf format specifiers are wrong

**Comment by rainers** [comment]
> Druntime for mingw printf format specifiers are wrong

The respective build is not done with `-de` so the compiler doesn't error out. `unittest.exe` is built successfully and is run, but crashes. IIRC DM make misinterprets the exit code and reports `not found`.

**Comment by adamdruppe** [comment]
Old bugzilla link came up on the forum:
https://issues.dlang.org/show_bug.cgi?id=7176

can perhaps use that for the changelog.

**Comment by CyberShadow** [comment]
> You need a preview switch for this.

I thought we only need preview switches for breaking changes. Why would strict additions need a compiler switch?

**Comment by wilzbach** [comment]
> > You need a preview switch for this.
> 
> I thought we only need preview switches for breaking changes. Why would strict additions need a compiler switch?

With a -preview switch it can be merged right now and formally approved via a DIP later.

**Comment by 12345swordy** [comment]
I concur. This needs a preview switch.

**Comment by Imperatorn** [comment]
Just be sure to add documentation and examples so it's clear to users how to use it. Other than that, 10 x 👍

**Comment by andre2007** [comment]
@adamdruppe any chance you can continue with this pr (add the missing parts)? This pr is highly interesting.

**Comment by adamdruppe** [comment]
ummm i guess i can look at it later, maybe over the weekend. I don't see any real value in the preview switch but it is probably simple enough to do.

**Comment by thewilsonator** [comment]
> can perhaps use that for the changelog.

Seems good.

**Comment by PetarKirov** [comment]
Hey @adamdruppe any chance you can continue work on this? I for one am quite excited about it :P

**Comment by adamdruppe** [comment]
On Thu, Dec 17, 2020 at 01:48:39PM -0800, Petar Kirov wrote:
> Hey @adamdruppe any chance you can continue work on this? I for one am quite excited about it :P

it is on my list, just been very busy lately between day job, baby, and
house work.

the core of it already works though, all that is left here is the
administrative overhead.

**Comment by adamdruppe** [comment]
I added the preview switch.

Can't the changelog be generated from the commit message? I don't know how these things work.

**Comment by thewilsonator** [comment]
You need to make the commit message say "Fix issue 7176: ... description goes here ...".

**Comment by PetarKirov** [comment]
> Can't the changelog be generated from the commit message? I don't know how these things work.

The changelog for each project is generally generated based on two sources:
* Fixed Bugzilla bugs & enhancements (which are auto-closed if the commit message follows the [right format](https://github.com/dlang-bots/dlang-bot#automated-references) as mentioned by @dlang-bot above)
* Manually written changelog entries in this folder: https://github.com/dlang/dmd/tree/master/changelog

But we always use manually written changelog entries for more notable changes, such as this one.

**Comment by Imperatorn** [comment]
So, what's the status on this one now? Ready for merge? Anything left to be done?

**Comment by 12345swordy** [comment]
You need see the red tags, as there they are very impotent.

**Comment by thewilsonator** [comment]
```
Test 'compilable/previewhelp.d' failed: 
...
diff:
----
   =rvaluerefparam   enable rvalue arguments to ref parameters
   =nosharedaccess   disable access to shared memory objects
   =in               `in` on parameters means `scope const [ref]` and accepts rvalues
+  =shortenedMethods allow use of => for methods and top-level functions in addition to lambas
   =inclusiveincontracts 'in' contracts of overridden methods must be a superset of parent contract
----
```

File: test/compilable/shortened_methods.d
Comment (MoonlightSentinel) [inline test/compilable/shortened_methods.d:1]:
```suggestion
// REQUIRED_ARGS: -preview=shortenedMethods
```

**Review by thewilsonator — APPROVED** [review]

**Comment by adamdruppe** [comment]
@thewilsonator thanks for the help things like this are never high on my todo list

**Comment by Geod24** [comment]
🎉

**Comment by WalterBright** [comment]
@thewilsonator @geod24 This is a language change. Why is there no changelog and no spec pull? Both were called out with labels. There's no indication of what the new BNF grammar is.

**Comment by Geod24** [comment]
@WalterBright : I didn't pull, just got the notification, but those are good point. Even though there's an issue for it (so it appears in the changelog), this definitely warrants its oown changelog entry.
Regarding the spec PR, I usually ask for it ([and I did](https://github.com/dlang/dmd/pull/11833#issuecomment-706129071)) but most people tend to forget. At least this one is behind a `-preview` switch.

**Comment by WalterBright** [comment]
We have standards for how a PR should be done. I certainly do not advocate blindly follow those standards, but for deviations there ought to be a decent reason. I expect all those with Pull privileges to help ensure this. If a PR doesn't meet those standards, an offer to help the submitter would be helpful.

I'm especially concerned about this because of the many legitimate complaints that the spec is out of date. This problem should be getting better over time, not worse. I added the "Need Changelog" label which was removed without explanation. I know you requested the Spec change, but that was ignored.

@12345swordy wrote as well "You need see the red tags, as there they are very impotent." The typo is appropriate.

**Comment by adamdruppe** [comment]
I don't speak spec and my understanding* is the rebase means the changelog picked it up. But maybe I'll get around to figuring them out but it will be several months, if ever. So if you want those kind of things done, it is far, far, far more efficient for someone who already knows it to do that part. Specialization of labor isn't just for building horseless carriages.

I'm not terribly invested in this PR. I just wrote a patch to prove to someone in chat it wasn't hard to implement. It isn't something I'll likely use myself since I avoid new features anyway (my libs promise compatibility with 3+ year old compilers...)


* BTW this readme
https://github.com/dlang/dmd/tree/master/changelog

should be linked from this page https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md

**Comment by Imperatorn** [comment]
While I'd like this to be merged, which it seems it did, we also have to follow some standards. Sometimes there are multiple forces pulling in different directions. At those times, I actually think Walter should have the final word (if it's not completely insane). Let's respect our founder(s)

**Comment by schveiguy** [comment]
Just seeing that this was merged. For 2.096.0, this will be in the compiler, it needs a spec udpate and a better changelog entry.

**Comment by 12345swordy** [comment]
This shouldn't be merge without a spec PR have been made and merged. If the next compiler release contain this feature without a spec update, it will create unnecessarily confusion for new users.

**Comment by Imperatorn** [comment]
Who can update the spec? One week until release 👀

**Comment by Geod24** [comment]
Someone needs to make a PR. And this is why we ask for spec PRs before. Once the feature is merged, there's less incentive to do so.

**Comment by Imperatorn** [comment]
@adamdruppe I think this might fall on you to do. If you cannot do it yourself, provide an alternative solution to how it could/should be done and what people should/have to be involved.

**Comment by schveiguy** [comment]
https://github.com/dlang/dmd/pull/12241

**Pull request merged**
